instr/gen_decode.py

```diff
@@ -1,14 +1,137 @@
+            write_buffer += f"m_imm = bit::signExtend<InstrCode ,{extend_from + 1}>(m_imm);\n"
+    
+    return write_buffer
+
+
+def GenerateSwitchCaseMasks(yaml_dump: dict) -> str:
+    write_buffer = ""
+    masks = dict(dict())
+    for inst in yaml_dump.get("instructions"):
+        if masks.get(inst.get("debug_hex_fixedmask")) == None:
+            masks[inst.get("debug_hex_fixedmask")] = {inst.get("mnemonic"):inst}
+        else:
+            masks[inst.get("debug_hex_fixedmask")][inst.get("mnemonic")] = inst
+
+            
+    for mask, inst_dict in masks.items():
+        write_buffer += f"switch (instr_code & 0x{mask}) {{\n"
+        for inst_name, inst in inst_dict.items():
+            match = inst.get("debug_hex_fixedvalue")
+
+            write_buffer += f"case 0x{match}:\n"
+            write_buffer += GenerateFieldAssigning(inst, yaml_dump.get("fields"))
+            write_buffer += "break;\n"
+
+        write_buffer += "default:\nbreak;\n}\n"
+
+
+    return write_buffer
+
+def GenerateHandleNode(node_dict : dict, field_dict: dict) -> str:
+    write_buffer = ""
+
+    for key, value in node_dict.items():
+
+        if type(key) is int:
+            write_buffer += f"case {key}:\n"
+            if value.get("range") is not None:
+                write_buffer += GenerateHandleNode(value, field_dict)
+            elif value.get("mnemonic") is not None:
+                write_buffer += GenerateFieldAssigning(value, field_dict)
+            
+            write_buffer += "break;\n"
+
+
+        elif key == "range":
+            #write_buffer += "std::cout << bit::getBitField<InstrCode, 6, 0>(instr_code) << std::endl;"
+            write_buffer += "switch ("
+            write_buffer += "bit::getBitField<InstrCode,"
+            write_buffer += f"{value.get('msb')},"
+            write_buffer += f"{value.get('lsb')}>(instr_code)"
+            write_buffer += ")\n"
+
+        elif key == "nodes":
+            write_buffer += "{\n"
+            write_buffer += GenerateHandleNode(value, field_dict)
+            write_buffer += "default:\nbreak;"
+            write_buffer += "}\n"
+
+        
+        else:
+            print("Unknown key:", key)
+
+    return write_buffer
+
+
+def GenerateDecoderTree(yaml_dump: dict) -> str:
+    write_buffer = ""
+    write_buffer += GenerateHandleNode(yaml_dump.get("decodertree"), yaml_dump.get("fields"))
+    
+
+    return write_buffer
+
+
+def main():
+    with open("risc-v.yaml") as f:
+        yaml_dump = dict(yaml.safe_load(f))
+
+    write_buffer = "#include <sim/instr.hpp>" "\n"  +\
+        "#include <sim/common.hpp> \n"              +\
+        "#include <iostream> \n"              +\
+        "\n"                                        +\
+        "namespace sim {" "\n"                      +\
+        "namespace instr {" "\n"                    +\
+        "\n"                                        +\
+        "Instr::Instr(InstrCode instr_code) {" "\n"
```

Comment:
Maybe, it will be more efficient to always assign rd, rs1 and rs2 here (and remove assigning from decode tree). This will reduce code size.

Reply:
Such changes require perf tests. Thus, i advice to deal with this later

Reply:
ok, might address later